src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,78 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
-/**
- * Make this class immutable. See requirements in the README file
- */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car implements Cloneable {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = getWheelsListCopy(wheels);
+        this.engine = (engine != null) ? new Engine(engine) : null;
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car addWheel(Wheel newWheel) {
+        ArrayList<Wheel> newListOfWheels = new ArrayList<>(this.wheels);
+        newListOfWheels.add(newWheel);
+        return new Car(year, color, newListOfWheels, engine);
+    }
+
+    public Car changeColor(String newColor) {
+        return new Car(year, newColor, wheels, engine);
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return getWheelsListCopy(wheels);
+    }
+
+    private List<Wheel> getWheelsListCopy(List<Wheel> originalList) {
+        List<Wheel> copy = new ArrayList<>();
```

Comment:
```suggestion
        List<Wheel> copy = new ArrayList<>(originalList.size());
```
if you know the capacity will be more than 10, it's better to set it from the beginning